src/commonMain/kotlin/org/kbson/serialization/Serializers.kt

```diff
@@ -0,0 +1,799 @@
+import kotlinx.serialization.json.JsonNull
+import kotlinx.serialization.json.JsonObject
+import kotlinx.serialization.json.JsonPrimitive
+import kotlinx.serialization.json.booleanOrNull
+import kotlinx.serialization.json.contentOrNull
+import kotlinx.serialization.json.doubleOrNull
+import kotlinx.serialization.json.floatOrNull
+import kotlinx.serialization.json.intOrNull
+import kotlinx.serialization.json.longOrNull
+import kotlinx.serialization.modules.SerializersModule
+import org.kbson.BsonArray
+import org.kbson.BsonBinary
+import org.kbson.BsonBoolean
+import org.kbson.BsonDBPointer
+import org.kbson.BsonDateTime
+import org.kbson.BsonDecimal128
+import org.kbson.BsonDocument
+import org.kbson.BsonDouble
+import org.kbson.BsonInt32
+import org.kbson.BsonInt64
+import org.kbson.BsonJavaScript
+import org.kbson.BsonJavaScriptWithScope
+import org.kbson.BsonMaxKey
+import org.kbson.BsonMinKey
+import org.kbson.BsonNull
+import org.kbson.BsonObjectId
+import org.kbson.BsonRegularExpression
+import org.kbson.BsonSerializationException
+import org.kbson.BsonString
+import org.kbson.BsonSymbol
+import org.kbson.BsonTimestamp
+import org.kbson.BsonType
+import org.kbson.BsonUndefined
+import org.kbson.BsonValue
+import org.kbson.internal.Base64Utils
+import org.kbson.internal.HexUtils
+import org.kbson.internal.validateSerialization
+
+/**
+ * The BsonSerializersModule
+ *
+ * Contains serializers that handle the conversion of Bson data to extended json and back
+ */
+public val BsonSerializersModule: SerializersModule = SerializersModule {
+    contextual(BsonValue::class, BsonValueSerializer)
+}
+
+
+/**
+ * The Bson companion object
+ */
+public object Bson {
+
+    private val extendedJson = Json { BsonSerializersModule }
+
+    /**
+     * Create a BsonDocument from a Json string
+     *
+     * @param json the Json String
+     * @return a BsonDocument
+     */
+    public operator fun invoke(jsonString: String): BsonValue {
+        return extendedJson.decodeFromString(BsonValueSerializer, jsonString)
+    }
+
+    /**
+     * Create a Json string from a BsonValue
+     *
+     * @param bsonValue the BsonValue
+     * @return the Json String
+     */
+    public fun toJson(bsonValue: BsonValue): String {
+        return extendedJson.encodeToString(BsonValueSerializer, bsonValue)
+    }
+}
+
+@OptIn(ExperimentalSerializationApi::class)
+@Serializer(forClass = BsonValue::class)
+private object BsonValueSerializer : KSerializer<BsonValue> {
+
+    override val descriptor: SerialDescriptor = buildClassSerialDescriptor("BsonDocument") {}
+
+    @Suppress("ComplexMethod")
+    override fun serialize(encoder: Encoder, value: BsonValue) {
+        when (encoder) {
+            is JsonEncoder -> {
+                when (value.bsonType) {
+                    BsonType.ARRAY -> ListSerializer(BsonValueSerializer).serialize(encoder, value.asArray())
+                    BsonType.DOCUMENT -> BsonDocumentSerializer.serialize(encoder, value.asDocument())
```

Comment:
Can this be just:

`MapSerializer(BsonDocumentKeySerializer, BsonValueSerializer).serialize(encoder, value.asDocument())`

similar to what's above for `ARRAY`?

Reply:
It could be but I still need a `BsonDocumentSerializer` for the scope field in Javascript with scope - so chose to reuse it here.